src/SignalR/server/Core/src/Internal/HubCallerClients.cs

```diff
@@ -104,4 +104,38 @@ public Task SendCoreAsync(string method, object?[] args, CancellationToken cance
             return _proxy.SendCoreAsync(method, args, cancellationToken);
         }
     }
+
+    private sealed class SingleClientProxy : ISingleClientProxy
+    {
+        private readonly ISingleClientProxy _proxy;
+        private readonly SemaphoreSlim _parallelInvokes;
+
+        public SingleClientProxy(ISingleClientProxy hubClients, SemaphoreSlim parallelInvokes)
+        {
+            _proxy = hubClients;
+            _parallelInvokes = parallelInvokes;
+        }
+
+        public async Task<T> InvokeCoreAsync<T>(string method, object?[] args, CancellationToken cancellationToken = default)
+        {
+            // Releases the SemaphoreSlim that is blocking pending invokes, which in turn can block the receive loop.
+            // Because we are waiting for a result from the client we need to let the receive loop run otherwise we'll be blocked forever
+            _parallelInvokes.Release();
+            try
+            {
+                var result = await _proxy.InvokeCoreAsync<T>(method, args, cancellationToken);
+                return result;
+            }
+            finally
+            {
+                // Re-acquire the SemaphoreSlim, this is because when the hub method completes it will call release
+                await _parallelInvokes.WaitAsync(CancellationToken.None);
```

Comment:
Just an idea:

Must `_parallelInvokes` be a `SemaphoreSlim` or can other sync-types be used?
E.g. with a bounded-channel (of `maxInvokeLimit`) the reader / writer pair can be used for synchronization too. IIf "waiting" will succeed synchronous, then the advantage is that it's a ValueTask, thus saving the allocation for the Task.
(Of course, if it real async mostly, then the benefit is much lower).

Reply:
Channels would completely break the concept I'm trying to implement here, so while it might be better from a Task alloc perspective, it is much worse from a behavior perspective 😞 

If we decide to go a different direction, then I think the Channel is worth revisiting to see if we can replace SemaphoreSlim.

Reply:
Maybe we talked past each other.

I meant something like this
```c#
using System.Threading.Channels;

ChannelBasedSemaphore semaphore = new(1);

ValueTask task0 = semaphore.WaitAsync();
ValueTask task1 = semaphore.WaitAsync();

Console.WriteLine(task0.IsCompleted);   // true
Console.WriteLine(task1.IsCompleted);   // false

semaphore.Release();

await Task.Yield();
Console.WriteLine(task1.IsCompleted);   // true

internal class ChannelBasedSemaphore
{
    private readonly Channel<int> _channel;

    public ChannelBasedSemaphore(int initialCount)
    {
        _channel = Channel.CreateBounded<int>(initialCount);

        for (int i = 0; i < initialCount; ++i)
        {
            _channel.Writer.TryWrite(42);    // any dummy-value will do
        }
    }

    public void Release() => _channel.Writer.TryWrite(42);

    public async ValueTask WaitAsync(CancellationToken cancellationToken = default)
    {
        _ = await _channel.Reader.ReadAsync(cancellationToken);
    }
}
```
as replacement for `SemaphoreSlim` (kind of drop-in replacement).
(Note: only hacked togehter, no perf-optimization, etc.).